src/App.tsx

```diff
@@ -1,26 +1,304 @@
+        }
+
+        setLoadingTodos(current => current.filter(id => id !== todo.id));
+      });
+  };
+
+  const handleAllToggleStatus = () => {
+    if (activeTodos.length > 0) {
+      activeTodos.forEach(handleToggleStatus);
+    } else {
+      completedTodos.forEach(handleToggleStatus);
+    }
+  };
+
+  const handleRename = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo(todo)
+      .then(updatedTodo =>
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+          ),
+        ),
+      )
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setLoadingTodos(current => current.filter(id => id !== todo.id));
+      });
+  };
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {todos.length !== 0 && (
+            <button
+              type="button"
+              className={classNames('todoapp__toggle-all', {
+                active: activeTodos.length === 0,
+              })}
+              data-cy="ToggleAllButton"
+              onClick={handleAllToggleStatus}
+            />
+          )}
+
+          <form onSubmit={event => handleAddTodo(event)}>
+            <input
+              ref={inputRef}
+              data-cy="NewTodoField"
+              type="text"
+              className="todoapp__new-todo"
+              placeholder="What needs to be done?"
+              value={title}
+              onChange={event => setTitle(event.target.value.trimStart())}
+            />
+          </form>
+        </header>
+
+        {todos.length > 0 && (
+          <>
+            <section className="todoapp__main" data-cy="TodoList">
+              {filterTodos.map(todo => (
+                <TodoItem
+                  key={todo.id}
+                  todo={todo}
+                  onDelete={() => handleDeleteTodo(todo.id)}
+                  onToggle={() => handleToggleStatus(todo)}
+                  onRename={handleRename}
+                  isLoading={loadingTodos.includes(todo.id)}
+                />
+              ))}
+
+              {tempTodo && (
+                <TodoItem
+                  todo={tempTodo}
+                  isLoading={loadingTodos.includes(0)}
```

Comment:
we always want show loader for tempTodo
```suggestion
                  isLoading
```
So you can also avoid adding `0` to loadingTodos (BTW, there are only ids, not todos, so better name is `loadingTodoIds`)